### PR TITLE
tests/build system: allow buildtests to be executed according to MCU groups

### PR DESCRIPTION
**Pull request by phiros**
The idea behind this PR is to group buildtests by MCU type, thus allowing us to take advantage of the buildmatrix feature of travis (might also be useful if you want to 
execute a buildtest locally just for a certain group of boards).

An example test run can be seen here:
https://travis-ci.org/phiros/RIOT/builds/49693219
EDIT: also the test run on this PR ;-)

The whole thing is still very hacky but might be a reasonable short term solution until we have our own test cluster.
Also I feel like travis is generally very unreliably when it comes to jobs which take more than 10 min...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,11 @@ language: c
...
 
 env:
-    - NPROC_MAX=8
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=msp430
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=arm7
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m0
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m4
 
 before_install:
```
Comment (authmillenon) [inline .travis.yml]:
```diff
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=x86
```
?

Comment (gebart) [inline .travis.yml]: and avr?

**Comment by gebart** [comment]
This is a good idea. Is it possible to reduce the number of packages to install on Travis initialization as well? Only install the toolchain for the targeted platform, I mean.

**Comment by LudwigOrtmann** [comment]
@gebart good idea about optimizing installations, but maybe a good candidate for a follow up PR once this one is finished.

**Comment by OlegHahm** [comment]
:+1: for this PR, for @gebart's proposal and for making it a follow-up PR.

**Comment by phiros** [comment]
I have incorporated most of the feedback I got:
* The packages are now selectively installed according the MCU group (saves a lot of time)
* Unit tests are run only in the x86 group for now (unit tests currently run on x86 anyways)

In addition to that I split up a larger group of MCUs into two (cortex_m3 -> cortex_m3 and cortex_m3_2) in order to keep the run time of each build task low.

File: boards/redbee-econotag/Makefile.features [hunk 2]
```diff
@@ -1 +1,3 @@
 FEATURES_PROVIDED += transceiver
+FEATURES_MCU_GROUP = arm7
+
```
Comment (LudwigOrtmann) [inline boards/redbee-econotag/Makefile.features]: stray blank line

File: .travis.yml [hunk 3]
```diff
@@ -3,7 +3,14 @@ language: c
...
 
 env:
-    - NPROC_MAX=8
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=avr8
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=msp430
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=x86
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=arm7
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m0
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3
+    - NPROC_MAX=8 BUILDTEST_MCU_GROUP=cortex_m3_2
```
Comment (LudwigOrtmann) [inline .travis.yml]: please cortex_m3 => cortex_m3_1 for consistency

**Comment by LudwigOrtmann** [comment]
Please rebase and check if any new boards popped up in master in the meantime.

**Comment by phiros** [comment]
rebased (no new boards were added in the meantime) + fixed those two issues you mentioned.

**Comment by gebart** [comment]
Is the cortex m3 split arbitrary or did you have any particular thoughts for which board went in which group? (just curious)

File: dist/tools/travis-scripts/build_and_test.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		./dist/tools/doccheck/check.sh master || exit
+
+		./dist/tools/externc/check.sh master || exit
+
+		# TODO:
+		#   Remove all but `master` parameters to cppcheck (and remove second
+		#   invocation) once all warnings of cppcheck have been taken care of
+		#   in master.
+		./dist/tools/cppcheck/check.sh master --diff-filter=MR --error-exitcode=0 || exit
+		./dist/tools/cppcheck/check.sh master --diff-filter=AC || exit
```
Comment (LudwigOrtmann) [inline dist/tools/travis-scripts/build_and_test.sh]: The static analysis does not really depend on "$BUILDTEST_MCU_GROUP" == "x86", maybe you could factor it into either it's own group, or have it executed in a different fashion?

File: dist/tools/travis-scripts/build_and_test.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -e
+if [[ $BUILDTEST_MCU_GROUP ]]
+then
+	if [ "$BUILDTEST_MCU_GROUP" == "x86" ]
+	then
+		make -s -C ./examples/default info-concurrency
+		git rebase riot/master || git rebase --abort
```
Comment (LudwigOrtmann) [inline dist/tools/travis-scripts/build_and_test.sh]: This is not specific to `x86`, should be done for all builds.

**Comment by LudwigOrtmann** [comment]
@phiros ping

**Comment by LudwigOrtmann** [comment]
Please rebase on master, there's a new script (from https://github.com/RIOT-OS/RIOT/pull/2468).

**Comment by phiros** [comment]
@gebart the split for the cortex m3 boards is arbitrary.

@LudwigOrtmann The static checks execute now for every MCU group (IMHO the overhead is tolerable; each MCU group build takes ~1 min. longer to finish). 
Do I get an ACK from you for this? I think we can improve on this further with follow up PRs if we really
see the need for it.

**Comment by LudwigOrtmann** [comment]
please squash

**Comment by LudwigOrtmann** [comment]
(ACK for the content)

**Comment by LudwigOrtmann** [comment]
I guess you want to remove the WIP label? You sound eager to merge this ;)

**Comment by LudwigOrtmann** [comment]
Don't forget to remove the 'needs squashing' label once you squashed.

**Comment by OlegHahm** [comment]
:raised_hands:

**Pull request merged**